### DISCUSSION TITLE
default-registry vs. registry in VCPKG_ROOT

### DISCUSSION THREAD

**Discussion by m-kuhn** [microsoft/vcpkg#36555] [Q&A]
Reading the documentation about registries, I am not completely sure how to interpret `"default-registry"` and `VCPKG_ROOT`.
E.g. https://learn.microsoft.com/en-us/vcpkg/consume/git-registries

Context:
So far (without any registries) I have added some `FetchContent` cmake logic to get a specific commit from the vcpkg repository and run some bootstrapping code and run in manifest mode. This allows isolated and well defined behavior for what I can see by only using `VCPKG_ROOT` with the default registry provided implicitly.

I am now adding a new registry (via `vcpkg-configuration.registries`) and am informed that I need to also specify a `default-registry` (or set it to `null`). I can obviously add the `microsoft/vcpkg` registry (again) but that means that I will have to keep the `baseline` (commit-ish) updated in two places and will have two git clones happening. With `null` the registry from `VCPKG_ROOT` is completely ignored.

I assume with registries specified through `vcpkg-configuration` the only thing that's used from `VCPKG_ROOT` i the `vcpkg` executable and the complete checkout that I'm doing through `FetchContent` is a quite heavy operation only to bootstrap that.

I could also have a vcpkg base installation (`VCPKG_ROOT`) initialized manually outside cmake. This would add a tiny bit of overhead for each contributor and would risk having an outdated vcpkg executable compared to what's defined in `default-registry`, I am unsure to what level this would be problematic.

Is my understanding correct and are there some best practices to handle this?

**Comment by m-kuhn** [comment]
I now realize that [a thing called `vcpkg-init` exists](https://github.com/microsoft/vcpkg-tool/blob/main/README.md#installuseremove), that's likely what I was missing. Still wondering if that's something that should be expected to be manually installed or if there's a "smart" way to initialize this from within cmake in a platform independent way.

**Comment by Osyotr** [comment]
Not pretending to give a best practice, but I use this setup:
- `vcpkg.json`: only list of dependencies
- `vcpkg-configuration.json`: builtin registry with some baseline as default
- vcpkg is checked out at baseline of a default registry (i.e. builtin)

Some time ago I wrote a cmake script that does that:
https://gist.github.com/Osyotr/e2d415312a10183bc7c614013494bf21

**Reply by m-kuhn** [reply]
That looks nice, but also results in cloning the complete baseline registry twice.

I wonder if the same could be achieved only for the tool/running vcpkg-init in a platform agnostic way. E.g. releasing a tar (or mini git repo) with vcpkg-init scripts for all platforms. That could FetchContent'ed or cloned and then initialize the tool only.

Any reason for putting the vcpkg-configuration into a separate file?

**Reply by m-kuhn** [reply]
I just had some experience that is discouraging me from this approach:

Given: a `vcpkg.json` with 

```
  "vcpkg-configuration": {
    "default-registry": {
      "kind": "git",
      "baseline": "80403036a665cb8fcc1a1b3e17593d20b03b2489",
      "repository": "https://github.com/microsoft/vcpkg"
    },
    "registries": [
      {
        "kind": "git",
        "baseline": "c288fd01cbc95c56a122747c3c39cbab2710ee17",
        "repository": "https://github.com/open-vcpkg/vcpkg-qt5",
        "packages": [
          "qscintilla-qt5",
          "qt5-webkit"
        ]
      }
    ]
  }
```

when using an external vcpkg installation version 2024-01-11 libgcrypt will fail with an include not found (`gpg-error.h`), when checking out the main vcpkg repo, bootstrapping that and using that, it will build fine.
I don't yet understand what's going on and can just express a momentary confusion.

**Reply by Osyotr** [reply]
Can't tell without any logs for analysis.
